numpy/core/src/private/mem_overlap.c

```diff
@@ -0,0 +1,814 @@
+/*
+  Solving memory overlap integer programs and bounded Diophantine equations with
+  positive coefficients.
+
+  Asking whether two strided arrays `a` and `b` overlap is equivalent to
+  asking whether there is a solution to the following problem::
+
+      sum(stride_a[i] * x_a[i] for i in range(ndim_a))
+      -
+      sum(stride_b[i] * x_b[i] for i in range(ndim_b))
+      ==
+      base_b - base_a
+
+      0 <= x_a[i] <= shape_a[i]
+      0 <= x_b[i] <= shape_b[i]
+
+  for some integer x_a, x_b.  Itemsize needs to be considered as an additional
+  dimension with stride 1 and size itemsize-1.
```

Comment:
I don't understand where the -1 comes from. I guess it's correct if we use the non-strict equality above, but in that case I think we'd need -1's on them too, and really it's easy just to use the half-open convention consistently? Or maybe I'm missing something.